demos/minimap/minimap.js

```diff
@@ -0,0 +1,247 @@
+};
+
+/**
+* Updates the overall position of the viewport of the minimap by appropriately
+* using translate functions.
+*/
+Minimap.scaleMinimap = function(){
+  var minimapBoundingBox = Minimap.minimap.getBlocksBoundingBox();
+  var workspaceBoundingBox = Minimap.workspace.getBlocksBoundingBox();
+  var workspaceMetrics = Minimap.workspace.getMetrics();
+  var minimapMetrics = Minimap.minimap.getMetrics();
+
+  //scaling the mimimap such that all the blocks can be seen in the viewport
+  //this padding is default because this is how to scrollbar(in main workspace) is implemented
+  var topPadding = (workspaceMetrics.viewHeight)*Minimap.minimap.scale/(2*Minimap.workspace.scale);
+
+  // if actual padding is more than half view ports height, change it to actual padding/
+  if((workspaceBoundingBox.y*Minimap.workspace.scale-workspaceMetrics.contentTop)
+      *Minimap.minimap.scale/Minimap.workspace.scale>topPadding){
+    topPadding=(workspaceBoundingBox.y*Minimap.workspace.scale-workspaceMetrics.contentTop)
+    *Minimap.minimap.scale/Minimap.workspace.scale;
+  }
+  var scalex = (minimapMetrics.viewWidth-Minimap.PADDING)/minimapBoundingBox.width;
+  var scaley = (minimapMetrics.viewHeight-2*topPadding)/minimapBoundingBox.height;
+  Minimap.minimap.setScale(Math.min(scalex,scaley));
+
+  // translating the minimap
+  Minimap.minimap.translate(-minimapBoundingBox.x*Minimap.minimap.scale+Minimap.PADDING,
+                          -minimapMetrics.contentTop*Minimap.minimap.scale + topPadding);
+};
+
+/**
+* Handles the onclick event on the minimapBoundingBox. Changes mapDraggers position.
+* @param {Event} e Event from the mouse click
+*/
+Minimap.updateMapDragger = function(e){
+  var y=e.clientY;
+  var draggerHeight = Minimap.mapDragger.getAttribute("height");
+  var finalY = y-Minimap.rect.top-draggerHeight/2;
+  var maxValidY = (Minimap.workspace.getMetrics().contentHeight-Minimap.workspace.getMetrics().viewHeight)
+                  *Minimap.minimap.scale;
+  if(y+draggerHeight/2>Minimap.rect.bottom){
+    finalY = Minimap.rect.bottom-Minimap.rect.top-draggerHeight;
+  }else if(y<Minimap.rect.top+draggerHeight/2){
+    finalY = 0;
+  }
+  // do not go below loew bound of scrollbar
+  if(finalY>maxValidY){
+    finalY=maxValidY;
+  }
+  Minimap.mapDragger.setAttribute("y", finalY);
+
+  // required, otherwise creates a feedback loop
+  Minimap.disableScrollChange=true;
+  Minimap.workspace.scrollbar.vScroll.set((finalY*Minimap.workspace.scale)/Minimap.minimap.scale);
+  Minimap.disableScrollChange=false;
+};
+
+/**
+* Handles the onclick event on the minimapBoundingBox, paramaters are passed by
+* the event handler
+* @param {Float} position This is the absolute postion of the scrollbar.
+* @param {boolean} horizontal Informs if the change event if for horizontal(true)
+*     scrollbar or vertical(false) scrollbar.
+*/
+Minimap.onScrollChange = function(position,horizontal){
+  if(horizontal){
```

Comment:
Hi Karan,

A few notes on why 2d navigation is important to us:
- Not all implementations of Blockly use vertical code. [Scratch Blocks](https://github.com/llk/scratch-blocks) explicitly includes a horizontal UI.
- Kids tend to organize their code in very different ways because they didn't start from text programming. We need to support them however they build.
- While a single piece of code grows down, it's often easier to write separate functions side by side instead of above/below each other so you can compare them more easily.

The [minimap used in Starcraft](http://wiki.teamliquid.net/commons/images/a/aa/SCminimap.png) is maybe a good example of what we were thinking of. It shows you where you are within the whole space and the bounding box accurately reflects what you see on screen.